### ISSUE TITLE
Error pushing huge packages on IIS (413 Payload Too Large)

### ISSUE THREAD

**Issue by Voxel7** [loic-sharma/BaGet#463]
I'm unable to push huge packages on BaGet installed on IIS.
It seems related to #88.

I didn't figure out by trial and error the maximum allowable size. I'm able to reproduce the bug with packages of at least 41mb.

Push command:
```batch
C:\> nuget.exe push -source http://myserver:5000/v3/index.json MyHugePackage.nupkg
AVERTISSEMENT : No API Key was provided and no API Key could be found for 'http://myserver:5000/api/v2/package'. To save an API Key for a source use the 'setApiKey' command.
Pushing MyHugePackage.nupkg to 'http://myserver:5000/api/v2/package'...
  PUT http://myserver:5000/api/v2/package/
  RequestEntityTooLarge http://myserver:5000/api/v2/package/ 25948ms
Le code d'état de réponse n'indique pas la réussite : 413 (Payload Too Large).
```

Output from the log file:
```log
fail: Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer[3]

      Unexpected exception in "IISHttpContext.ReadBody".

Microsoft.AspNetCore.Server.IIS.BadHttpRequestException: Request body too large.

   at Microsoft.AspNetCore.Server.IIS.BadHttpRequestException.Throw(RequestRejectionReason reason)

   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadBody()

fail: Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer[2]

      Connection ID "18086456109425492159", Request ID "800000c0-0001-fb00-b63f-84710c7967bb": An unhandled exception was thrown by the application.

Microsoft.AspNetCore.Server.IIS.BadHttpRequestException: Request body too large.

   at Microsoft.AspNetCore.Server.IIS.BadHttpRequestException.Throw(RequestRejectionReason reason)

   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadBody()

   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()

   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)

   at System.IO.Pipelines.Pipe.GetReadAsyncResult()

   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadAsync(Memory`1 memory, CancellationToken cancellationToken)

   at Microsoft.AspNetCore.Server.IIS.Core.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)

   at Microsoft.AspNetCore.WebUtilities.BufferedReadStream.EnsureBufferedAsync(Int32 minCount, CancellationToken cancellationToken)

   at Microsoft.AspNetCore.WebUtilities.MultipartReaderStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)

   at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)

   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)

   at Microsoft.AspNetCore.Http.Features.FormFeature.InnerReadFormAsync(CancellationToken cancellationToken)

   at Microsoft.AspNetCore.Mvc.ModelBinding.FormValueProviderFactory.AddValueProviderAsync(ValueProviderFactoryContext context)

   at Microsoft.AspNetCore.Mvc.ModelBinding.CompositeValueProvider.CreateAsync(ActionContext actionContext, IList`1 factories)

   at Microsoft.AspNetCore.Mvc.ModelBinding.CompositeValueProvider.TryCreateAsync(ActionContext actionContext, IList`1 factories)

   at Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()

--- End of stack trace from previous location where exception was thrown ---

   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)

   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)

   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)

   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)

   at BaGet.OperationCancelledMiddleware.Invoke(HttpContext context) in D:\a\1\s\src\BaGet.Core.Server\OperationCancelledMiddleware.cs:line 28

   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()
```

I installed BaGet on IIS as detailed in the documentation ([Windows IIS Proxy](https://loic-sharma.github.io/BaGet/quickstart/iis-proxy/))

I also increase all the size limits I found:

web.config
```xml
<system.webServer>
  <security>
    <requestFiltering>
      <requestLimits maxAllowedContentLength="2147483647" maxQueryString="4096" />
    </requestFiltering>
  </security>
</system.webServer>
```
```xml
<system.web>
  <httpRuntime maxRequestLength="2147483647" />
</system.web>
```
```xml
<system.serviceModel>
  <bindings>
    <basicHttpBinding>
      <binding maxBufferPoolSize="2147483647" maxReceivedMessageSize="2147483647" maxBufferSize="2147483647" transferMode="Streamed">
        <readerQuotas maxBytesPerRead="2147483647" maxStringContentLength="2147483647‬" />
      </binding>	
    </basicHttpBinding>
  </bindings>	
</system.serviceModel>
```

applicationHost.config
```xml
<location path="BaGet server">
  <system.webServer>
    <serverRuntime maxRequestEntityAllowed="2147483647" uploadReadAheadSize="2147483647" />
  </system.webServer>
</location>
```

**Comment by RichardSinden** [comment]
It's limited to 30MB (https://github.com/aspnet/Announcements/issues/267). I was having exactly the same issue and I couldn't find any way to change this in configuration without a code change. 
So I've created this PR to make it configurable in the appsettings.json.

**Comment by softlion** [comment]
If I remember well there is an environment variable to set that does the same thing, and no need to set it in code.

**Comment by RichardSinden** [comment]
@softlion, I don't suppose you happen to know what it is? I can't find anything that is supposed to change it beyond what @Voxel7 tried above.

**Comment by Ugenx** [comment]
> @softlion, I don't suppose you happen to know what it is? I can't find anything that is supposed to change it beyond what @Voxel7 tried above.

Would this be what you're looking for?

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-2.2#kestrel-options-1

The available properties for that config setting are available here: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.core.kestrelserverlimits?view=aspnetcore-3.1

You should be able to either control this via an appSettings file or define an environment variable that maps to this property structure, i.e. Kestrel__Limits__MaxRequestBodySize

**Comment by RichardSinden** [comment]
Thanks for the suggestion, but these are the Kestrel config settings and the code is already configured to have an unlimited max-size when hosted in Kestrel. @loic-sharma suggested that hardcoding this was fine for Kestrel as anyone requiring a max-size could host in IIS. IIS seems to ignore all the usual web.config settings that people suggest for configuring max request size, and I can't find any reference to an environment variable for this. So I don't see any alternative to a code change to allow it to be configurable, which, incidentally, is ready to be merged in - **nudge**, **nudge**, @loic-sharma .

**Comment by softlion** [comment]
IIS is only used as a reverse proxy for netcore apps.
You can change the setting in IIS config.
The same apply for kestrel, if hosted behind nginx for example.

![Increasing the Maximum Upload File Size in IIS - KB306 - Inf - Google Chrome](https://user-images.githubusercontent.com/190756/74930679-77719900-53de-11ea-851e-8d749887c785.jpg)

**Comment by RichardSinden** [comment]
This setting seems to match what is configured in `<requestFiltering>` in the web.config, as above, but it doesn't seem to do anything in this case.
```xml
<system.webServer>
  <security>
    <requestFiltering>
      <requestLimits maxAllowedContentLength="2147483647" maxQueryString="4096" />
    </requestFiltering>
  </security>
</system.webServer>
```

**Comment by Voxel7** [comment]
@softlion What you propose is equivalent to changing `requestLimits` from `web.config`, but using the user interface. Normally, it should have solved the problem, but not for BaGet.

The pull request #473 addresses the problem well (thanks @RichardSinden!). I'll just wait for the merge.

**Comment by reduckted** [comment]
Does the fix from #473 still work for anyone? It looks like it may have been broken by the refactoring in #538. The `IISServerOptions` are no longer configured anywhere, which means (as far as I can tell), putting this in the appsettings file has no effect:

```json
"IISServerOptions": {
    "MaxRequestBodySize": 262144000
}
```

@loic-sharma

**Comment by reduckted** [comment]
I've found and fixed the problem in #653 and #654. The default `MaxRequestBodySize` was always overwriting the value specified in appsettings.
